### PR TITLE
Issue 201: Test cleanup

### LINKED ISSUE #201: Test cleanups
# Task Description

We need to go over all the tests and remove all the test methods that are there just for the sake of code coverage to pass. Here is an example:
```java
    /**
     * http://stackoverflow.com/questions/9700179/junit-testing-helper-class-with-only-static-methods
     */
    @Test
    public void just_to_silence_coverage()
    {
        new AttributesUtils()
        {
            // ignore this
        };
    }
```

We should go over the rest of the classes and simply remove such tests, as they are of no value and are increasing the code coverage just for the sake of appearances, which is extremely bad. 

As part of this cleanup, all test case methods that are using underscores shoulc be converted to camelCase (check the example above, but there may be others).

# Tasks

The following tasks will need to be carried out:
* [ ] Drop any test methods that are simply there for the sake of skewing up the code coverage numbers.
* [ ] Go over the tests and make sure that all the methods in them are camelCase and make any changes that may be required.

# Help

* Our [chat channel](https://chat.carlspring.org/channel/s3fs-nio-community)
* Points of contact:
  * @carlspring
  * @ptirador
  * @steve-todorov
  * @sbespalov

### PR DESCRIPTION
**Pull request by nmonokov**
# Pull Request Description

This pull request closes #201 .

# Acceptance Test

* [ ] Building the code with `mvn clean install -Punit-tests,it-minio,it-s3` still works.

# Questions

* Does this pull request break backward compatibility? 
  * [ ] Yes
  * [x] No

* Does this pull request require other pull requests to be merged first? 
  * [ ] Yes, please see #...
  * [x] No

* Does this require an update of the documentation?
  * [ ] Yes, please see [provide details here]
  * [x] No

### CODE REVIEW

**Comment by carlspring** [comment]
@nmonokov ,

Could you please clarify for each of the removed test methods why you think they should be removed? I have the feeling it's a step too much than what is required and would therefore appreciate the explanation. Indeed, there is a lot to be desired from both the codebase and the tests.

**Comment by nmonokov** [comment]
@carlspring waiting for your decision about `GetPathTest` test. And is it possible to run `mvn clean install -Punit-tests,it-minio,it-s3` on local machine. 'Cause I see integration tests and I don't know how to verify them except for running them on some stage env.

**Comment by nmonokov** [comment]
> @nmonokov ,
> 
> Could you please clarify for each of the removed test methods why you think they should be removed? I have the feeling it's a step too much than what is required and would therefore appreciate the explanation. Indeed, there is a lot to be desired from both the codebase and the tests.

sure. I'll do it.

File: src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/NewAsynchronousFileChannelTest.java
Comment (steve-todorov) [inline src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/NewAsynchronousFileChannelTest.java:102]: :+1: this is essentially the same test as `fileChannelSize` but with bigger `content` (bigger in terms of the first one is `content` and the second one is `content-more-large` lol :D)

:+1: for `fileChannelTruncate` as well (same thing)

File: src/test/java/org/carlspring/cloud/storage/s3fs/spike/InstallProviderTest.java
Comment (steve-todorov) [inline src/test/java/org/carlspring/cloud/storage/s3fs/spike/InstallProviderTest.java:9]: Please revert this change since nothing else actually changed in this test. :)

Comment (nmonokov) [reply src/test/java/org/carlspring/cloud/storage/s3fs/spike/InstallProviderTest.java:9]: oh, I'm really sorry for that. I haven't tuned my IDE correctly.

File: src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/NewAsynchronousFileChannelTest.java
Comment (nmonokov) [reply src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/NewAsynchronousFileChannelTest.java:102]: yeah, i've removed a lot of duplicates mostly. Tests from 13-17 years has a lot of those.

File: src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/SetAttributeTest.java
Comment (steve-todorov) [inline src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/SetAttributeTest.java:48]: :+1: for removing this dead code, but you forgot to remove the no longer used imports:

```
import org.carlspring.cloud.storage.s3fs.S3FileSystem;
import org.carlspring.cloud.storage.s3fs.S3UnitTestBase;
import org.carlspring.cloud.storage.s3fs.util.S3EndpointConstant;

import java.io.IOException;
import java.nio.file.FileSystemNotFoundException;
import java.nio.file.FileSystems;
```

should become

```
import org.carlspring.cloud.storage.s3fs.S3UnitTestBase;
import org.carlspring.cloud.storage.s3fs.util.S3EndpointConstant;

import java.io.IOException;
```

File: src/test/java/org/carlspring/cloud/storage/s3fs/path/S3PathTest.java
Comment (steve-todorov) [inline src/test/java/org/carlspring/cloud/storage/s3fs/path/S3PathTest.java:129]: This method could be renamed to `registerWatcherShouldThrowException` which would make it clear what's the intended result.

```suggestion
    /**
     * This covers {@link S3Path#register (WatchService watcher, WatchEvent.Kind<?>... events))
     */
    @Test
    void registerWatcherShouldThrowUnsupportedOperationException()
```

File: src/test/java/org/carlspring/cloud/storage/s3fs/path/S3PathTest.java
Comment (steve-todorov) [inline src/test/java/org/carlspring/cloud/storage/s3fs/path/S3PathTest.java:129]: :+1:

File: src/test/java/org/carlspring/cloud/storage/s3fs/path/S3PathTest.java
Comment (steve-todorov) [inline src/test/java/org/carlspring/cloud/storage/s3fs/path/S3PathTest.java:151]: This should be reverted - it appears to have been intended to cover the case when you have all 3 arguments in the `.register()`, but was implemented incorrectly. The proper implementation would be:

```

    /**
     * This covers {@link S3Path#register (WatchService watcher, WatchEvent.Kind<?>[] events, WatchEvent.Modifier... modifiers)
     */
    @Test
    void registerWatcherWithEventsAndModifierShouldThrowUnsupportedOperationException()
    {
        // We're expecting an exception here to be thrown
        Exception exception = assertThrows(UnsupportedOperationException.class, () -> {
            forPath("file1").register(null, null, null);
        });

        assertNotNull(exception);
    }
```

File: src/test/java/org/carlspring/cloud/storage/s3fs/spike/ProviderSpecTest.java
Comment (steve-todorov) [inline src/test/java/org/carlspring/cloud/storage/s3fs/spike/ProviderSpecTest.java:56]: We might need this test despite it being marked as `FIXME`. There is a difference between `StandardOpenOption.CREATE, StandardOpenOption.WRITE, StandardOpenOption.READ` in `seekable()` and  `StandardOpenOption.APPEND` in `seekableRead()`.

File: src/test/java/org/carlspring/cloud/storage/s3fs/spike/InstallProviderTest.java
Comment (steve-todorov) [reply src/test/java/org/carlspring/cloud/storage/s3fs/spike/InstallProviderTest.java:9]: No problem! :)

**Comment by steve-todorov** [comment]
> And is it possible to run mvn clean install -Punit-tests,it-minio,it-s3 on local machine. 'Cause I see integration tests and I don't know how to verify them except for running them on some stage env.

@nmonokov yes, it's possible to run it locally. You need:

1. Docker (for `it-minio`)
2. Valid S3 Credentials (check the [dev guide](https://s3fs-nio.carlspring.org/contributing/developer-guide/#s3-advanced))

Then you just `mvn clean install -Punit-tests,it-minio,it-s3` and it should run all of them.

FYI, you can skip the `it-minio` profile right now since we only have the dummy tests running there at the moment.

**Comment by nmonokov** [comment]
@steve-todorov I don't understand why run-it-s3 is failing. I haven't done any changes to those tests and they run locally. Can you help me understand what is wrong here and how do I fix it please?

**Comment by steve-todorov** [comment]
TL;DR If they run locally against S3 - ignore GH Actions failure.

The long version - this is a known GH Actions issue which currently have no real solution. Secrets are not available to forks unless the user is part of the organization. If you fork this project and setup the env vars in your fork - it'll build there. Other than building in your fork or locally there isn't anything you or we can do.

File: src/test/java/org/carlspring/cloud/storage/s3fs/S3FileAttributesTest.java
Comment (nmonokov) [inline src/test/java/org/carlspring/cloud/storage/s3fs/S3FileAttributesTest.java:42]: This test is a duplicate to toStringPrintsBasicInfo

File: src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/NewAsynchronousFileChannelTest.java
Comment (nmonokov) [inline src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/NewAsynchronousFileChannelTest.java:281]: duplicate to fileChannelTruncate

File: src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/NewByteChannelTest.java
Comment (nmonokov) [inline src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/NewByteChannelTest.java:94]: duplicate to seekableSize

File: src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/NewByteChannelTest.java
Comment (nmonokov) [inline src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/NewByteChannelTest.java:292]: duplicato to seekableTruncate

File: src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/NewDirectoryStreamTest.java
Comment (nmonokov) [inline src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/NewDirectoryStreamTest.java:57]: duplicate to createStreamDirectoryReader

File: src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/NewDirectoryStreamTest.java
Comment (nmonokov) [inline src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/NewDirectoryStreamTest.java:152]: duplicate to list999Paths

File: src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/NewInputStreamTest.java
Comment (nmonokov) [inline src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/NewInputStreamTest.java:62]: duplicate to inputStreamFile

File: src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/NewOutputStreamTest.java
Comment (nmonokov) [inline src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/NewOutputStreamTest.java:198]: duplicate to outputStreamWithCreateAndFileNotExists

File: src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/ReadAttributesTest.java
Comment (nmonokov) [inline src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/ReadAttributesTest.java:140]: This test is pointless due to existence of readAttributesDirectory

File: src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/ReadAttributesTest.java
Comment (nmonokov) [inline src/test/java/org/carlspring/cloud/storage/s3fs/fileSystemProvider/ReadAttributesTest.java:407]: duplicate to readPosixPermissionGroupExecuteAttributes

**Comment by nmonokov** [comment]
@carlspring I've added all comments to deleted tests. As you see, they are mostly duplicates.
@steve-todorov so, if build fail is a github issue, what should i do next?

**Comment by carlspring** [comment]
Would it make sense to make any of the duplicate tests as `@Parameterized` ones?

**Comment by nmonokov** [comment]
Most of them are just exact duplicates. Only two or something like that are with different values. I'm talking about list999Paths and list1050Paths. If value before and after 1000 is crucial, then yes, we can do this test as parametrized.

**Comment by steve-todorov** [comment]
Yeah, most of the removed tests are redundant - having 2 bytes more in the `file content` and called `anotherTest` :D 
Not entirely sure why those were added to begin with. :) 

> @steve-todorov so, if build fail is a github issue, what should i do next?

Have you ran over all of the tests? If there isn't anything else you can find then we could merge this PR after you've signed the ICLA and mailed it to @carlspring. :)

**Comment by nmonokov** [comment]
@steve-todorov could you elaborate on signing this ICLA please? :)

**Comment by steve-todorov** [comment]
@nmonokov https://s3fs-nio.carlspring.org/contributing/legal/ :)

**Comment by carlspring** [comment]
Hi @nmonokov ,

As explained in the page @steve-todorov provided, you will have to sign the [ICLA](https://s3fs-nio.carlspring.org/assets/resources/pdfs/ICLA.pdf) and mail it back to me.

Kind regards,

Martin

**Comment by nmonokov** [comment]
Hey, @carlspring! I've sent signed ICLA.

**Comment by carlspring** [comment]
Thanks for signing the ICLA! :)

**Comment by nmonokov** [comment]
@carlspring @steve-todorov so to avoid those build fails I should just make branch per feature without forking?

**Comment by steve-todorov** [comment]
> so to avoid those build fails I should just make branch per feature without forking?

You simply **cannot** avoid the build failure - it's a github limitation which causes the failure. :)
If you decide to work on future issues - then you should continue creating feature branches i.e. `issues/1234`.

**Review by steve-todorov — APPROVED** [review]
LGTM! Thanks for your work on this! :)

**Comment by carlspring** [comment]
Thanks for your help with this, @nmonokov and congratulations on your first pull request on the project! :)

**Pull request merged**
